File: 프로그래머스/3/132266. 부대복귀/부대복귀.java
import java.util.*;
class Solution {
    //다익스트라
    int dist[];

    public int[] solution(int n, int[][] roads, int[] sources, int destination) {
        int[] answer = new int[sources.length];
        ArrayList<Integer> arr[];
        arr=new ArrayList[n+1];
        dist=new int[n+1];
        for(int i=1;i<=n;i++){
            arr[i]=new ArrayList<>();
        }
        Arrays.fill(dist, 987654321);
        dist[destination]=0;
        
        int m=roads.length;
        for(int i=0;i<m;i++){
            int st = roads[i][0];
            int en = roads[i][1];
            arr[st].add(en);
            arr[en].add(st);
        }
        PriorityQueue<Integer> pq = new PriorityQueue<>();
        pq.add(destination);
        while(!pq.isEmpty()){
            int cur = pq.poll();
            for(int nxt:arr[cur]){
                if(dist[nxt]>dist[cur]+1){
                    dist[nxt]=dist[cur]+1;
                    pq.add(nxt);
                }
            }
            
        }
        for(int i=0;i<sources.length;i++){
            answer[i]=dist[sources[i]];
            if(answer[i]==987654321) answer[i]=-1;
        }
        return answer;
    }
}

File: 프로그래머스/3/132266. 부대복귀/부대복귀.cpp
#include <string>
#include <vector>
#include <queue>
#include <iostream>

using namespace std;
vector<int> v[100001];
priority_queue<pair<int, int>, vector<pair<int, int>>, greater<>> pq; // {이동거리, nxt}
int dist[100001];
vector<int> solution(int n, vector<vector<int>> roads, vector<int> sources, int destination) {
    vector<int> answer;
    fill(dist, dist+n+1, 987654321);
    for(int i=0;i<roads.size();i++){
        int a = roads[i][0];
        int b = roads[i][1];
        
        v[a].push_back(b);
        v[b].push_back(a);
    }
    pq.push({0, destination});
    dist[destination]=0;
    while(!pq.empty()){
        int cur = pq.top().second;
        int d = pq.top().first;
        
        pq.pop();

        if(dist[cur]<d) continue;
        for(auto nxt:v[cur]){
            if(d+1<dist[nxt]){
                dist[nxt]=d+1;
                pq.push({dist[nxt], nxt});
            }
        }
    }
    for(auto i: sources){
        if(dist[i]==987654321) dist[i]=-1;
        answer.push_back(dist[i]);
    }
    return answer;
}